File: app/src/main/java/com/example/vladimir/easyenglishlearn/exercises/WordConstructorViewModel.java
package com.example.vladimir.easyenglishlearn.exercises;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.databinding.ObservableField;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static com.example.vladimir.easyenglishlearn.Constants.EMPTY_STRING;

public class WordConstructorViewModel extends ExerciseViewModel {

    private MutableLiveData<List<Character>> mCharArrayLiveData;
    private List<Character> mLetterList;
    public final ObservableField<String> answer = new ObservableField<>();


    public WordConstructorViewModel() {
        mLetterList = new ArrayList<>();
        mCharArrayLiveData = new MutableLiveData<>();
    }

    void onNewButtonClick(String letter) {
        mAnswerBuilder.append(letter);
        answer.set(mAnswerBuilder.toString());
        mLetterList.remove(Character.valueOf(letter.charAt(0)));
        if (mLetterList.isEmpty()) {
            checkAnswer();
        }
    }

    public void onButtonUndoClick() {
        if (mAnswerBuilder.length() > 0) {
            int lastCharIndex = mAnswerBuilder.length() - 1;
            char letterFromButton = mAnswerBuilder.charAt(lastCharIndex);
            mAnswerBuilder.deleteCharAt(lastCharIndex);
            answer.set(mAnswerBuilder.toString());
            mLetterList.add(letterFromButton);
            mCharArrayLiveData.setValue(mLetterList);
        }
    }

    @Override
    void prepareQuestionAndAnswers() {
        super.prepareQuestionAndAnswers();

        splitCurrentWord();
        Collections.shuffle(mLetterList);
        mCharArrayLiveData.setValue(mLetterList);
    }

    @Override
    void cleanPreviousData() {
        super.cleanPreviousData();
        mLetterList.clear();
        answer.set(EMPTY_STRING);
    }

    private void splitCurrentWord() {
        char[] letters = mTranslationDirection
                ? mCurrentWord.getTranslation().toCharArray()
                : mCurrentWord.getLexeme().toCharArray();
        for (char letter : letters) {
            mLetterList.add(letter);
        }
    }

    LiveData<List<Character>> getCharArrayLiveData() {
        return mCharArrayLiveData;
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/exercises/WordConstructorViewModel.kt
package com.vladimir_x.easyenglishlearn.exercises

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.vladimir_x.easyenglishlearn.Constants
import java.util.*

class WordConstructorViewModel : ExerciseViewModel() {
    private val _charArrayLiveData = MutableLiveData<List<Char>>()
    private val letterList: MutableList<Char> = ArrayList()
    var answer: String = ""

    val charArrayLiveData: LiveData<List<Char>>
        get() = _charArrayLiveData

    fun onNewButtonClick(letter: String) {
        answerBuilder.append(letter)
        answer = answerBuilder.toString()
        letterList.remove(Character.valueOf(letter[0]))
        if (letterList.isEmpty()) {
            checkAnswer()
        } else {
            _charArrayLiveData.value = letterList
        }
    }

    fun onButtonUndoClick() {
        if (answerBuilder.isNotEmpty()) {
            val lastCharIndex = answerBuilder.lastIndex
            val letterFromButton = answerBuilder[lastCharIndex]
            answerBuilder.deleteCharAt(lastCharIndex)
            answer = answerBuilder.toString()
            letterList.add(letterFromButton)
            _charArrayLiveData.value = letterList
        }
    }

    override fun prepareQuestionAndAnswers() {
        super.prepareQuestionAndAnswers()
        splitCurrentWord()
        letterList.shuffle()
        _charArrayLiveData.value = letterList
    }

    override fun cleanPreviousData() {
        super.cleanPreviousData()
        letterList.clear()
        answer = Constants.EMPTY_STRING
    }

    private fun splitCurrentWord() {
        val letters = if (translationDirection) {
            currentWord?.translation?.toCharArray()
        } else {
            currentWord?.lexeme?.toCharArray()
        }
        letters?.let {
            for (letter in letters) {
                letterList.add(letter)
            }
        }
    }
}